### PR TITLE
[5.1] Fix SEF redirect while executing parseRule

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

SEF plugin should not do redirect while parsing the URL.
Original PR:
- https://github.com/joomla/joomla-cms/pull/42702
- https://github.com/joomla/joomla-cms/pull/42704

### Testing Instructions
Please re-test linked PRs, they should still work as before.

### Actual result BEFORE applying this Pull Request
Works

### Expected result AFTER applying this Pull Request
Works

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed
- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

Ping @Hackwar

### CODE REVIEW

File: plugins/system/sef/src/Extension/Sef.php [hunk 1]
```diff
@@ -322,38 +325,30 @@ public function addTrailingSlash(&$router, &$uri)
...
+    protected function enforceTrailingSlash()
     {
-        // We only want to redirect on GET requests
-        if ($this->app->getInput()->getMethod() != 'GET') {
-            return;
-        }
-
         $originalUri = Uri::getInstance();
 
         if ($this->params->get('trailingslash') == 1 && substr($originalUri->getPath(), -1) == '/' && $originalUri->toString() != Uri::root()) {
```
Comment (Quy) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if ($this->params->get('trailingslash') == 1 && str_ends_with($originalUri->getPath(), '/') && $originalUri->toString() != Uri::root()) {
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 2]
```diff
@@ -322,38 +325,30 @@ public function addTrailingSlash(&$router, &$uri)
...
             // Remove trailingslash
             $originalUri->setPath(substr($originalUri->getPath(), 0, -1));
-            $this->app->redirect($originalUri->toString(), 301);
+            $this->getApplication()->redirect($originalUri->toString(), 301);
         } elseif ($this->params->get('trailingslash') == 2 && substr($originalUri->getPath(), -1) != '/') {
```
Comment (Quy) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        } elseif ($this->params->get('trailingslash') == 2 && !str_ends_with($originalUri->getPath(), '/')) {
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 3]
```diff
@@ -322,38 +325,30 @@ public function addTrailingSlash(&$router, &$uri)
...
     {
-        // We only want to redirect on GET requests
-        if ($this->app->getInput()->getMethod() != 'GET') {
-            return;
-        }
-
         $originalUri = Uri::getInstance();
 
-        if ($this->params->get('trailingslash') == 1 && substr($originalUri->getPath(), -1) == '/' && $originalUri->toString() != Uri::root()) {
+        if ($this->params->get('trailingslash') == 1 && str_ends_with($originalUri->getPath(), '/') && $originalUri->toString() != Uri::root()) {
```
Comment (HLeithner) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        if ((int)$this->params->get('trailingslash') === 1 && str_ends_with($originalUri->getPath(), '/') && $originalUri->toString() !== Uri::root()) {
```

File: plugins/system/sef/src/Extension/Sef.php [hunk 4]
```diff
@@ -322,38 +325,30 @@ public function addTrailingSlash(&$router, &$uri)
...
             // Remove trailingslash
             $originalUri->setPath(substr($originalUri->getPath(), 0, -1));
-            $this->app->redirect($originalUri->toString(), 301);
-        } elseif ($this->params->get('trailingslash') == 2 && substr($originalUri->getPath(), -1) != '/') {
+            $this->getApplication()->redirect($originalUri->toString(), 301);
+        } elseif ($this->params->get('trailingslash') == 2 && !str_ends_with($originalUri->getPath(), '/')) {
```
Comment (HLeithner) [inline plugins/system/sef/src/Extension/Sef.php]:
```suggestion
        } elseif ((int)$this->params->get('trailingslash') === 2 && !str_ends_with($originalUri->getPath(), '/')) {
```

**Review by HLeithner — COMMENTED** [review]
Didn't we talked about a redirection queue somewhere?

**Review by Hackwar — APPROVED** [review]

**Comment by SniperSister** [comment]
I have tested this item :white_check_mark: successfully on 35f024e9497370ce82e6185434ed9e5524a6cd4f<br /><br />I've tested the changes, works as described!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42932">issues.joomla.org/tracker/joomla-cms/42932</a>.</sub>

**Comment by Fedik** [comment]
> Didn't we talked about a redirection queue somewhere?

Do not remember anything about it

**Comment by bembelimen** [comment]
Thx

**Pull request merged**